flink-core/src/main/java/org/apache/flink/core/classloading/ComponentClassLoader.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.classloading;
+
+import org.apache.flink.util.function.FunctionWithException;
+
+import org.apache.flink.shaded.guava18.com.google.common.collect.Iterators;
+
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.Iterator;
+
+/**
+ * A {@link URLClassLoader} that restricts which classes can be loaded to those contained within the
+ * given classpath, except classes from a given set of packages that are either loaded parent or
+ * child-first.
+ */
+public class ComponentClassLoader extends URLClassLoader {
+    private static final ClassLoader PLATFORM_OR_BOOTSTRAP_LOADER;
+
+    private final ClassLoader parentClassLoader;
+
+    private final String[] parentFirstPackages;
+    private final String[] childFirstPackages;
```

Comment:
I am not sure whether we should call these fields parent and child because it looks a bit more like two independent class loaders (Flink and plugin, left and right, owner and component).

Reply:
yes I had similar thoughts, but thought that child/parent are the names people are most familiar with. When it comes to classloading I usually only differentiate between Flink and plugins etc, and leave out the bootstrap CL.

Owner and component is a good idea.